gax/src/main/java/com/google/api/gax/batching/v2/BatchingDescriptor.java

```diff
@@ -0,0 +1,95 @@
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching.v2;
+
+import com.google.api.core.SettableApiFuture;
+import java.util.List;
+
+/**
+ * An adapter that packs and unpacks the elements in and out of batch requests and responses.
+ *
+ * <p>This interface should be implemented by either a service specific client or autogenerated by
+ * gapic-generator.
+ *
+ * <p>Example implementation:
+ *
+ * <pre>{@code
+ * class ListBatchingDescriptor implements BatchingDescriptor<String, String, List<String>,
+ * List<String>> {
+ *
+ *   RequestBuilder<String, List<String>> newRequestBuilder(List<String> prototype) {
+ *     return new RequestBuilder<String, List<String>>() {
+ *       List<String> list = prototype.clone();
+ *
+ *       void add(String element) {
+ *         list.add(element);
+ *       }
+ *
+ *       List<String> build() {
+ *         return list.clone();
+ *       }
+ *     };
+ *   }
+ *
+ *   void splitResponse(List<String> batchResponse, List<SettableApiFuture<String>> batch) {
+ *     for (int i = 0; i < batchResponse.size(); i++) {
+ *       batch.get(i).set(batchResponse.get(i);
+ *     }
+ *   }
+ *
+ *   void splitException(Throwable throwable, List<SettableApiFuture<String>> batch) {
+ *     for (SettableApiFuture<String> result : batch) {
+ *       result.setException(throwable);
+ *     }
+ *   }
+ *
+ *   long countBytes(String element) {
+ *     return element.length();
+ *   }
+ * }
+ * }</pre>
+ */
+public interface BatchingDescriptor<ElementT, ResultT, RequestT, ResponseT> {
```

Comment:
Will it grouping element-request, result-response make it interface cleaner:
`<ElementT, RequestT, ResultT, ResponseT>`? 

Also, `ResultT` and `ResponseT` are effectively synonyms in English, and look very confusing when put together in same context (my first impression was: "What's the difference?"). 

Can we rename some of them so it is cleaner from the name that one is a repeated field in another and it is also consistent with the other pair? For example:
`<RequestElementT, RequestT, ResponseElementT, ResponseT>`

Reply:
The descriptor acts like an adapter between the element types and the batch types. I think the grouping of element types then batch types makes more sense. So as a compromise I propose:

`<ElementT, ElementResultT, RequestT, ResponseT>`

Reply:
@vam-google Please let us know your thoughts on the proposed type.

Reply:
@igorbernstein2 @rahulKQL Ok